src/reader.rs

```diff
@@ -142,63 +176,64 @@ pin_project! {
 /// [`futures::Stream`]: https://docs.rs/futures/0.3/futures/stream/trait.Stream.html
 /// [`MuxedEvents`]: struct.MuxedEvents.html
 /// [`LineSet`]: struct.LineSet.html
+#[derive(Debug)]
 pub struct MuxedLines {
     #[pin]
     events: crate::MuxedEvents,
-    reader_positions: HashMap<PathBuf, u64>,
-    readers: HashMap<PathBuf, LineReader>,
-    pending_readers: HashSet<PathBuf>,
+    inner: Option<Inner>,
     stream_state: StreamState,
 }
 }
 
 impl MuxedLines {
-    pub fn new() -> Self {
-        MuxedLines {
-            events: crate::MuxedEvents::new(),
-            reader_positions: HashMap::new(),
-            readers: HashMap::new(),
-            pending_readers: HashSet::new(),
+    pub fn new() -> io::Result<Self> {
+        Ok(MuxedLines {
+            events: crate::MuxedEvents::new()?,
+            inner: Some(Inner::new()),
             stream_state: StreamState::default(),
-        }
+        })
     }
 
     fn reader_exists(&self, path: &PathBuf) -> bool {
         // Make sure there isn't already a reader for the file
-        self.readers.contains_key(path) || self.pending_readers.contains(path)
+        self.inner.as_ref().unwrap().reader_exists(path)
     }
 
     /// Adds a given file to the lines watch, allowing for files which do not
     /// yet exist.
     ///
     /// Returns the canonicalized version of the path originally supplied, to
     /// match against the one contained in each `LineSet` received. Otherwise
-    /// returns `Error` for a given registration failure.
+    /// returns `io::Error` for a given registration failure.
     pub async fn add_file(&mut self, path: impl Into<PathBuf>) -> io::Result<PathBuf> {
         let source = path.into();
 
-        let source = self
-            .events
-            .add_file(&source)
-            .map_err(|e| io::Error::new(io::ErrorKind::AlreadyExists, format!("{:?}", e)))?;
+        let source = self.events.add_file(&source)?;
 
         if self.reader_exists(&source) {
             return Ok(source);
         }
 
         if !source.exists() {
-            let didnt_exist = self.pending_readers.insert(source.clone());
+            let didnt_exist = self.inner.as_mut().unwrap().insert_pending(source.clone());
 
             // If this fails it's a bug
             assert!(didnt_exist);
         } else {
-            let size = std_metadata(&source)?.len();
+            let size = metadata(&source).await?.len();
 
-            let reader = new_linereader(&source, Some(size))?;
+            let reader = new_linereader(&source, Some(size)).await?;
 
-            self.reader_positions.insert(source.clone(), size);
+            self.inner
+                .as_mut()
```

Comment:
Super minor, but if you add something like `let inner_mut = self.inner.as_mut().unwrap();`, you can simplify the next couple of lines to

```rust
let inner_mut = self.inner.as_mut().unwrap();
inner_mut.insert_reader_position(source.clone(), size);
let last = inner_mut.insert_reader(source.clone(), reader);
```